File: LeetCodeJava/38.报数.java
/*
 * @lc app=leetcode.cn id=38 lang=java
 *
 * [38] 报数
 */
class Solution {
    public String countAndSay(int n) {
        String x = "1";
        for (int i = 0; i < n; i ++) {
            x = nextInt(x);
        }
        return x;
    }

    private String nextInt(String x) {
        String ret = "";
        char pre;
        char now;
        int cnt = 0;
        pre = ' ';
        for (int i = 0; i < x.length(); i ++) {
            now = x.charAt(i);
            if (pre == ' ') {
                pre = now;
                cnt = 1;
            } else if (now == pre) {
                cnt += 1;
            } else {
                ret = cnt + String.valueOf(pre) + ret;
                pre = now;
                cnt = 1;
            }
        }
        ret = cnt + pre + ret;
        return ret;
    }
}

File: LeetCodePython/38.报数.py
#
# @lc app=leetcode.cn id=38 lang=python3
#
# [38] 报数
#

# @lc code=start
class Solution:
    def countAndSay(self, n: int) -> str:
        ret = '1'
        for i in range(1, n):
            tmp = ''
            pre = None
            cnt = 0
            for char in ret:
                if pre == None or char == pre:
                    cnt += 1
                elif char != pre:
                    tmp += str(cnt) + pre
                    cnt = 1
                pre = char
            ret = tmp + str(cnt) + pre
        return ret
# @lc code=end

